### ISSUE TITLE
Labels URI encoding when passing to rTorrent

### ISSUE THREAD

**Issue by Hawk84r** [Novik/ruTorrent#1603]
When i stopped using AutoMove because of bug #1602 and switched to .rtorrent.rc method described [here](https://github.com/rakshasa/rtorrent/wiki/Common-Tasks-in-rTorrent#move-completed-torrents-to-a-fixed-location), i discovered that labels with spaces in them (like "some good stuff") would be passed to rtorrent as "some%20good%20stuff", thus moving them to correspondingly named directory. This happens because of encodeURIComponent(lbl) on line 2033 of webui.js.

**Comment by deed02392** [comment]
Yup, same issue here. #metoo

**Comment by Novik** [comment]
This is a normal behavior.
Because we can't send non-UTF characters via xmlrpc. But labels may contains such characters.

**Comment by Hawk84r** [comment]
How is normal space non-UTF? And how is it actually working for me without encoding? This should not be normal behavior.

**Comment by Novik** [comment]
> How is normal space non-UTF?

This is a strange question. Correct question sholud be "For which reason we need to use encoding?"
See comment above for answer.

> And how is it actually working for me without encoding?

Sorry, but I'm not write WebUI specially and only for you. As result, I must to see on the common case.

**Comment by Hawk84r** [comment]
> Sorry, but I'm not write WebUI specially and only for you. As result, I must to see on the common case.
So, i am the only one in the world using spaces, ruTorrent and intended config-like downloads moving? Lucky me.
 
> This is a strange question. Correct question sholud be "For which reason we need to use encoding?"
How about because in this particular case we don't? But yeah, easy way does it.

**Comment by deed02392** [comment]
@Novik The encoding for XMLRPC is not the problem. The problem is we can't use non-URI characters as a label (e.g. / and space) if we set them via ruTorrent. Is this a limitation of the rTorrent XMLRPC then? Because of course you can use non-URI characters if you set them in rTorrent directly.

**Comment by Novik** [comment]
> So, i am the only one in the world using spaces

rtorrent's variable 'custom_1' is not specially intended for storing labels. ruTorrent uses it by such way. And yes, ruTorrent correctly works with a labels with spaces etc.
If you want to use this variable outside ruTorrent - then this is *your* problem. Not mine.

> The problem is we can't use non-URI characters as a label (e.g. / and space) if we set them via ruTorrent. 

Really? See screenshot. Corresponding label contains '/' and space and ruTorrent works with it correctly.
![2017-12-22_16h17_24](https://user-images.githubusercontent.com/1230409/34299387-9eb5174a-e733-11e7-93f0-5a8741ea048b.png)

**Comment by deed02392** [comment]
@Novik Thanks for the quick reply. So, yes it looks fine in the web GUI that non-URI characters are set as the label. But actually what rTorrent is seeing is the URI encoded equivalent of that. So if you have an rTorrent script for example using the label to set a target `mv` directory, then rTorrent tries to do `mv %2fnew-folder` instead of `mv /new-folder`.

**Comment by Novik** [comment]
> So if you have an rTorrent script for example

1) And as I already told above, this is not ruTorrent problem. Right?
2) https://stackoverflow.com/questions/6250698/how-to-decode-url-encoded-string-in-shell

**Comment by deed02392** [comment]
@Novik Only if it's impossible to pass these characters to rTorrent via XMLRPC. If that is not possible then it is not an ruTorrent problem, but I thought you said XMLRPC supports all of UTF-8?

**Comment by Novik** [comment]
> If that is not possible then it is not an ruTorrent problem

This is not ruTorrent problem in any case. Because UI works with a labels correctly.

**Comment by deed02392** [comment]
OK I see what you are saying - we are making advanced use of the labels feature but as far as you are concerned, the label feature is just for organising within ruTorrent. I still think this is an issue but I can understand why you do not care.

I will see if this issue can be fixed, and make a PR if so.

**Comment by deed02392** [comment]
I figured it out.

ruTorrent performs double URI encoding of the labels, which is unnecessary. It is performed first in `webui.js` and again doing the `rTorrentStub` request generation (`rTorrentStub.prototype.getCommon = function(cmd)` cmd="setlabel").

This means that even though the action handler I linked earlier does URL decoding, the final XMLRPC request sent to rTorrent contains the URL encoded data.

We can fix this by not performing the URL encoding in webui.js, as @Hawk84r initially suggested.
